### PR TITLE
feat: implement process for aggregate instance

### PR DESCRIPTION
**Pull request by alresarena2021**
feat: implement process for aggregate instance

### CODE REVIEW

File: libs/node-client/src/aggregate-instance.ts [hunk 1]
```diff
@@ -47,7 +64,73 @@ export class AggregateInstance<
...
+
+    try {
+      let events = await this.client.listAggregateEvents({
+        aggregate: {
+          id: this._id,
+          version: this._version,
+        },
+      });
+
+      for (const event of events) {
```
Comment (rogermadjos) [inline libs/node-client/src/aggregate-instance.ts]: this code block is common to `process` and `reload`. let's move it to a separate method and call it `digest`

File: libs/node-client/src/aggregate-instance.ts [hunk 2]
```diff
@@ -47,7 +64,73 @@ export class AggregateInstance<
   }
 
   public async process(command: TCommand): Promise<void> {
-    throw new Error('not implemented');
+    const context = this;
+    this.backoff.on('ready', async function () {
+      try {
+        await context.processCommand(command);
+      } catch (err) {
```
Comment (rogermadjos) [inline libs/node-client/src/aggregate-instance.ts]: generate error when it's not `InvalidAggregateVersionError`

File: libs/node-client/src/aggregate-instance.ts [hunk 3]
```diff
@@ -47,7 +64,73 @@ export class AggregateInstance<
...
+        if (err instanceof InvalidAggregateVersionError)
+          context.backoff.backoff();
+      }
+    });
+
+    try {
```
Comment (rogermadjos) [inline libs/node-client/src/aggregate-instance.ts]: this is redundant, just call `#backoff()` to execute `#processCommand` immediately

**Review by rogermadjos — CHANGES_REQUESTED** [review]

File: libs/node-client/src/aggregate-instance.ts [hunk 4]
```diff
@@ -47,7 +64,73 @@ export class AggregateInstance<
...
+
+        this._state = state;
+        this._version = event.aggregate.version;
+      }
+
+      const commandHandler = this.commandHandlers.get(command.type);
+
+      assert(commandHandler, `command handler for ${command.type} not found`);
+
+      const generatedEvent = await commandHandler.handle(
```
Comment (rogermadjos) [inline libs/node-client/src/aggregate-instance.ts:130]: `generatedEvent` only contains `type` and `body`. you need to set the `id` and `aggregate_version`

**Review by rogermadjos — CHANGES_REQUESTED** [review]

File: libs/node-client/src/aggregate-instance.ts (hunk 3 above)
Comment (alresarena2021) [reply libs/node-client/src/aggregate-instance.ts:97]: the backoffInstance.on read runs on background, calling backoff function wont wait the processCommand to finish inside of backoffInstance.on ready and ends the process function call

File: libs/node-client/src/aggregate-instance.ts [hunk 5]
```diff
@@ -61,19 +114,74 @@ export class AggregateInstance<
...
+        {
+          state: this._state,
+        } as TContext & { state: TState },
+        command
+      );
 
-        this._state = state;
-        this._version = event.aggregate.version;
+      let lastEvent = null;
```
Comment (rogermadjos) [inline libs/node-client/src/aggregate-instance.ts]: the command handler may generate multiple events, so this should be an array

File: libs/node-client/src/aggregate-instance.ts [hunk 6]
```diff
@@ -61,19 +114,74 @@ export class AggregateInstance<
...
+
+      if (Array.isArray(generatedEvent)) {
+        for await (const event of generatedEvent) {
```
Comment (rogermadjos) [inline libs/node-client/src/aggregate-instance.ts]: if the command handler generates multiple events, the version should increment for every event

**Review by rogermadjos — CHANGES_REQUESTED** [review]

File: libs/node-client/src/aggregate-instance.ts (hunk 6 above)
Comment (alresarena2021) [reply libs/node-client/src/aggregate-instance.ts]: fixed

**Pull request merged**